internal/component/common/loki/client/manager_wal.go

```diff
@@ -16,132 +16,148 @@ import (
+		return nil, fmt.Errorf("at least one client config must be provided")
+	}
 
-// MarkerHandler re-defines the interface of internal.MarkerHandler that the queue client interacts with, to contribute
-// to the feedback loop of when data from a segment is read from the WAL, or delivered.
-type MarkerHandler interface {
-	UpdateReceivedData(segmentId, dataCount int) // Data queued for sending
-	UpdateSentData(segmentId, dataCount int)     // Data which was sent or given up on sending
-	Stop()
-}
+	writer, err := wal.NewWriter(walCfg, logger, reg)
+	if err != nil {
+		return nil, fmt.Errorf("error creating wal writer: %w", err)
+	}
 
-// queuedBatch is a batch specific to a tenant, that is considered ready to be sent.
-type queuedBatch struct {
-	TenantID string
-	Batch    *batch
+	m := &WALManager{
+		writer: writer,
+		pairs:  make([]pair, 0, len(clientCfgs)),
+	}
+
+	var (
+		metrics      = NewMetrics(reg)
+		clientsCheck = make(map[string]struct{})
+
+		walWatcherMetrics = wal.NewWatcherMetrics(reg)
+		walMarkerMetrics  = internal.NewMarkerMetrics(reg)
+		walClientMetrics  = NewWALClientMetrics(reg)
+	)
+
+	for _, cfg := range clientCfgs {
+		// Don't allow duplicate clients, we have client specific metrics that need at least one unique label value (name).
+		clientName := getClientName(cfg)
+		if _, ok := clientsCheck[clientName]; ok {
+			return nil, fmt.Errorf("duplicate client configs are not allowed, found duplicate for name: %s", cfg.Name)
+		}
+		clientsCheck[clientName] = struct{}{}
+
+		// add some context information for the logger the watcher uses
+		wlog := log.With(logger, "client", clientName)
+
+		markerFileHandler, err := internal.NewMarkerFileHandler(logger, walCfg.Dir)
+		if err != nil {
+			return nil, err
+		}
+		markerHandler := internal.NewMarkerHandler(markerFileHandler, walCfg.MaxSegmentAge, logger, walMarkerMetrics.WithCurriedId(clientName))
+
+		client, err := newWalClient(metrics, walClientMetrics.CurryWithId(clientName), cfg, logger, markerHandler)
+		if err != nil {
+			return nil, fmt.Errorf("error starting wal client: %w", err)
+		}
+
+		// subscribe watcher's wal.WriteTo to writer events. This will make the writer trigger the cleanup of the wal.WriteTo
+		// series cache whenever a segment is deleted.
+		writer.SubscribeCleanup(client)
+
+		watcher := wal.NewWatcher(walCfg.Dir, clientName, walWatcherMetrics, client, wlog, walCfg.WatchConfig, markerHandler)
+
+		// subscribe watcher to wal write events
+		writer.SubscribeWrite(watcher)
+
+		level.Debug(logger).Log("msg", "starting WAL watcher for client", "client", clientName)
+		watcher.Start()
+
+		m.pairs = append(m.pairs, pair{
+			watcher: watcher,
+			client:  client,
+		})
+	}
+
+	return m, nil
 }
 
-// queue wraps a buffered channel and a routine that reads from it, sending batches of entries.
-type queue struct {
-	client *queueClient
-	q      chan queuedBatch
-	quit   chan struct{}
-	wg     sync.WaitGroup
-	logger log.Logger
+type pair struct {
```

Comment:
We need a better name for this. This looks more like a placeholder name, it's not helping with readability, but I think it could

Reply:
Ah yes this was a placeholder name